### PR TITLE
Forgiveness library changes

### PR DESCRIPTION
**Pull request by kevin-wangzefeng**
**What this PR does / why we need it**:
Splited from #34825, contains library changes that are needed to implement forgiveness:

1. ~~make taints-tolerations matching respect timestamps, so that one toleration can just tolerate a taint for only a period of time.~~ As TaintManager is caching taints and observing taint changes, time-based checking is now outside the library (in TaintManager). see #40355.
2. make tolerations respect wildcard key.
3. add/refresh some related functions to wrap taints-tolerations operation.

**Which issue this PR fixes**: 
Related issue: #1574
Related PR: #34825, #39469 
~~Please note that the first 2 commits in this PR come from #39469 .~~

**Special notes for your reviewer**:

~~Since currently we have `pkg/api/helpers.go` and `pkg/api/v1/helpers.go`, there are some duplicated periods of code laying in these two files.~~

~~Ideally we should move taints-tolerations related functions into a separate package (pkg/util/taints), and make it a unified set of implementations. But I'd just suggest to do it in a follow-up PR after Forgiveness ones done, in case of feature Forgiveness getting blocked to long.~~

**Release note**:

```release-note
make tolerations respect wildcard key
```

### CODE REVIEW

**Comment by davidopp** [comment]
@gmarek

File: pkg/api/helpers.go
Comment (timothysc) [inline pkg/api/helpers.go]: Could you also please outline the expected behavior in the godoc section because some folks have been confused by the default behavior, and possibly link to the document.

File: pkg/api/helpers.go
Comment (timothysc) [inline pkg/api/helpers.go]: not a fan of the name isInterestingTaint vs applyFilter, or other examples that are similar in the code.

File: pkg/api/helpers_test.go
Comment (timothysc) [inline pkg/api/helpers_test.go]: This is an an example of where I think the name of isInterestingTaint can be confusing.  It's a filter apply and any person not familiar with the entire spec will be confused.

File: pkg/api/types.go
Comment (timothysc) [inline pkg/api/types.go]: Why not Duration or some other time based object b/c int64 is signed.?.?

File: pkg/api/v1/helpers.go
Comment (timothysc) [inline pkg/api/v1/helpers.go]: I'm always flummoxed by our api-machinery, is there a reason we dupe helpers -> v1/helpers?

**Review by timothysc — CHANGES_REQUESTED** [review]
Could you separate the generated code into a separate commit please.  Only minor comments and we'll need rejig verification on the conversion to fields. 

/cc @aveshagarwal @derekwaynecarr

File: pkg/api/helpers.go
Comment (gmarek) [inline pkg/api/helpers.go]: Assuming it'll still be be in annotations... @davidopp

File: pkg/api/helpers.go
Comment (gmarek) [inline pkg/api/helpers.go]: Plus this comment makes no sense in this line;)

File: pkg/api/helpers.go
Comment (gmarek) [inline pkg/api/helpers.go]: @timothysc - this is the real reason why it should be `time.Duration` :)

File: pkg/api/helpers.go
Comment (gmarek) [inline pkg/api/helpers.go]: I still don't like this convention and having ptr here, but I won't argue too hard.

File: pkg/api/helpers.go
Comment (gmarek) [inline pkg/api/helpers.go]: Why? I can imagine use case when user want e.g. to tolerate 'NoSchedule' taint only for some time, as after that time he really need given Pod running. @davidopp

File: pkg/api/helpers.go
Comment (gmarek) [inline pkg/api/helpers.go]: +1

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go
Comment (gmarek) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: remove

**Comment by gmarek** [comment]
Can we move various small changes (scheduler, kubectl, etc.) to separate PR?

**Comment by gmarek** [comment]
Reviewed 22 of 22 files at r1.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/api/helpers.go, line 503 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39914#-KaS_lkkbbrA7ohR4kuv-r1-503:-KaqAdJDl1x58E7qy8BB:b-51yize) ([raw file](https://github.com/kubernetes/kubernetes/blob/76087b742e86f2fe3af9c2bee1b4e8c9e8dbf79e/pkg/api/helpers.go#L503)):*
<details><summary><i>Previously, gmarek (Marek Grabowski) wrote…</i></summary><blockquote>

Assuming it'll still be be in annotations... @davidopp 
</blockquote></details>

I spoke with David and learned that promotion of Taints to beta will be done later.

---

*[pkg/api/helpers.go, line 544 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39914#-KaS_lkkbbrA7ohR4kuv-r1-544:-KaqAr-OJgFX21GHJYGi:b-pv89l9) ([raw file](https://github.com/kubernetes/kubernetes/blob/76087b742e86f2fe3af9c2bee1b4e8c9e8dbf79e/pkg/api/helpers.go#L544)):*
<details><summary><i>Previously, gmarek (Marek Grabowski) wrote…</i></summary><blockquote>

@timothysc - this is the real reason why it should be `time.Duration` :)
</blockquote></details>

Sadly we don't use it in our API objects...

---

*[pkg/api/types.go, line 1800 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39914#-KaS_lkkbbrA7ohR4kux-r1-1800:-KaqB4yEr-5dNTQbbLKq:b-k5poih) ([raw file](https://github.com/kubernetes/kubernetes/blob/76087b742e86f2fe3af9c2bee1b4e8c9e8dbf79e/pkg/api/types.go#L1800)):*
<details><summary><i>Previously, timothysc (Timothy St. Clair) wrote…</i></summary><blockquote>

Why not Duration or some other time based object b/c int64 is signed.?.?
</blockquote></details>

For some reason we're not using those in API objects. @bgrant0607  @erictune 

---

*[pkg/api/v1/helpers.go, line 331 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39914#-KaS_lkkbbrA7ohR4kv--r1-331:-KaqBF7wBL6cylvDac9V:bal0s8h) ([raw file](https://github.com/kubernetes/kubernetes/blob/76087b742e86f2fe3af9c2bee1b4e8c9e8dbf79e/pkg/api/v1/helpers.go#L331)):*
<details><summary><i>Previously, timothysc (Timothy St. Clair) wrote…</i></summary><blockquote>

I'm always flummoxed by our api-machinery, is there a reason we dupe helpers -> v1/helpers? 
</blockquote></details>

@lavalamp @smarterclayton 

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39914)*

File: pkg/api/helpers.go
Comment (gmarek) [inline pkg/api/helpers.go]: OK, I spoke with David. We're keeping it in annotations in this PR.

**Comment by bgrant0607** [comment]
Why not duration:
Consistency, convenience of non-Go clients, avoid need for a page of
documentation of syntax. See API conventions

On Jan 20, 2017 12:44 AM, "Marek Grabowski" <notifications@github.com>
wrote:

*@gmarek* commented on this pull request.
------------------------------

In pkg/api/helpers.go <https://github.com/kubernetes/kubernetes/pull/39914>:

> @@ -499,6 +499,10 @@ func GetTolerationsFromPodAnnotations(annotations map[string]string) ([]Tolerati
 	return tolerations, nil
 }

+func GetPodTolerations(pod *Pod) ([]Toleration, error) {
+	return GetTolerationsFromPodAnnotations(pod.Annotations)

OK, I spoke with David. We're keeping it in annotations in this PR.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/39914>, or mute the thread

**Comment by gmarek** [comment]
@kevin-wangzefeng - any progress on your PRs? I have prototype basing on those nearly ready.

**Comment by kevin-wangzefeng** [comment]
Review status: 0 of 20 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/api/v1/helpers.go, line 331 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39914#-KaS_lkkbbrA7ohR4kv--r1-331:-KbFUwlK-oedxMxUH6aw:bcyzyqh) ([raw file](https://github.com/kubernetes/kubernetes/blob/76087b742e86f2fe3af9c2bee1b4e8c9e8dbf79e/pkg/api/v1/helpers.go#L331)):*
<details><summary><i>Previously, gmarek (Marek Grabowski) wrote…</i></summary><blockquote>

@lavalamp @smarterclayton 
</blockquote></details>

It's in the mid term of api-machinery change that the PR initially sent, now this duplication no longer exists.

---

*[plugin/pkg/scheduler/algorithm/predicates/predicates.go, line 1161 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39914#-KaS_lkkbbrA7ohR4kvD-r1-1161:-KbFgh0qtweDvf8seMsy:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/76087b742e86f2fe3af9c2bee1b4e8c9e8dbf79e/plugin/pkg/scheduler/algorithm/predicates/predicates.go#L1161)):*
<details><summary><i>Previously, gmarek (Marek Grabowski) wrote…</i></summary><blockquote>

remove
</blockquote></details>

Done.

---

*[pkg/api/helpers.go, line 525 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39914#-KaS_lkkbbrA7ohR4kuv-r1-525:-KbFgsgndkud03ogB7NN:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/76087b742e86f2fe3af9c2bee1b4e8c9e8dbf79e/pkg/api/helpers.go#L525)):*
<details><summary><i>Previously, gmarek (Marek Grabowski) wrote…</i></summary><blockquote>

Plus this comment makes no sense in this line;)
</blockquote></details>

Done.

---

*[pkg/api/helpers.go, line 575 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39914#-KaS_lkkbbrA7ohR4kuv-r1-575:-KbFgpMaBGAwrXr3SQ23:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/76087b742e86f2fe3af9c2bee1b4e8c9e8dbf79e/pkg/api/helpers.go#L575)):*
<details><summary><i>Previously, gmarek (Marek Grabowski) wrote…</i></summary><blockquote>

+1
</blockquote></details>

Done.

---

*[pkg/api/helpers_test.go, line 555 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39914#-KaS_lkkbbrA7ohR4kuw-r1-555:-KbFgkFCiz_G4K5U546X:b63i66b) ([raw file](https://github.com/kubernetes/kubernetes/blob/76087b742e86f2fe3af9c2bee1b4e8c9e8dbf79e/pkg/api/helpers_test.go#L555)):*
<details><summary><i>Previously, timothysc (Timothy St. Clair) wrote…</i></summary><blockquote>

This is an an example of where I think the name of isInterestingTaint can be confusing.  It's a filter apply and any person not familiar with the entire spec will be confused. 
</blockquote></details>

updated the description

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39914)*

File: pkg/api/v1/helpers.go
Comment (kevin-wangzefeng) [inline pkg/api/v1/helpers.go]: It's in the mid term of api-machinery change that the PR initially sent, now this duplication no longer exists.

**Comment by kevin-wangzefeng** [comment]
@gmarek, refreshed, PTAL

**Review by timothysc — APPROVED** [review]

**Comment by gmarek** [comment]
@kevin-wangzefeng - those changes are mostly OK, but I'd rather have API changes merged first and be clear on semantics we want to have.

**Comment by kevin-wangzefeng** [comment]
@gmarek sure.

**Comment by gmarek** [comment]
Reviewed 13 of 22 files at r2, 4 of 4 files at r3, 10 of 10 files at r4, 4 of 6 files at r5, 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[pkg/api/v1/helpers.go, line 337 at r6](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39914#-KbUQ2_E1-JTDQheHdta:-KbUQ2_E1-JTDQheHdtb:b24tm8f) ([raw file](https://github.com/kubernetes/kubernetes/blob/bd20a10d0bfb173cb47513308266f01f8257157c/pkg/api/v1/helpers.go#L337)):*
> ```Go
> //     If toleration.key is empty, toleration.operator must be 'Exists';
> //     this combination means to match all taint values and all taint keys.
> // (4) Nil toleration.tolerationSeconds means to match taints with
> ```

`timeAdded` doesn't impact this.  nil tolerationSeconds means that we want to tolerate given taint(s) forever.

---

*[pkg/api/v1/helpers.go, line 340 at r6](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39914#-KbUQH1H9m7nRPP1mC_5:-KbUQH1IrnefO5RXencE:b-bqd7ze) ([raw file](https://github.com/kubernetes/kubernetes/blob/bd20a10d0bfb173cb47513308266f01f8257157c/pkg/api/v1/helpers.go#L340)):*
> ```Go
> //     any value of taint.timeAdded.
> // (5) Non-nil positive toleration.tolerationSeconds means to
> //     match the taint for only a duration that starts at taint.timeAdded.
> ```

since the taint was observed by the TaintManager (controller? - I don't know what's the current naming scheme)

---

*[pkg/api/v1/helpers.go, line 351 at r6](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39914#-KbUQ_Iuxbr1y6kFKU7H:-KbUQ_IvjVPCl2X9IP41:b-kqc8cv) ([raw file](https://github.com/kubernetes/kubernetes/blob/bd20a10d0bfb173cb47513308266f01f8257157c/pkg/api/v1/helpers.go#L351)):*
> ```Go
> 
> 	if t.TolerationSeconds != nil {
> 		// taint with no timeAdded indicated can only be tolerated
> ```

Honestly - this is annoying (especially when writing tests). Plus it's duplicated functionality of `Effect` check. We have two options here:
- keep the logic as it's currently written, which means that there will be no way of specifying with single Toleration logic:
"tolerate everything, but `NoExecute` Taints only for 10 minutes"
- remove this check (i.e. tolerate `NoSchedule` taints  even when toleration has TolerationSeconds set, the disadvantage is that we'd need to make sure that for taints other than `NoExecute` this field is ignored and make sure that all users know about this + it's easy to find.

I'm strongly supporting the latter. @davidopp @timothysc 

---

*[pkg/api/v1/helpers.go, line 389 at r6](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39914#-KbUSHbYDGsMMzfvXL3L:-KbUSHbYDGsMMzfvXL3M:b-dale14) ([raw file](https://github.com/kubernetes/kubernetes/blob/bd20a10d0bfb173cb47513308266f01f8257157c/pkg/api/v1/helpers.go#L389)):*
> ```Go
> // TolerationsTolerateTaintsWithFilter checks if given tolerations tolerates
> // all the taints that apply to the filter in given taint list.
> func TolerationsTolerateTaintsWithFilter(tolerations []Toleration, taints []Taint, applyFilter taintsFilterFunc) bool {
> ```

Is this function used anywhere?

---

*[pkg/api/v1/types.go, line 2065 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39914#-KbUT47ad6qxT-hgGaD8:-KbUT47bkQrK5qWExC2Z:biql52t) ([raw file](https://github.com/kubernetes/kubernetes/blob/af77f44f089e74f8f4456929b7765e14375b1861/pkg/api/v1/types.go#L2065)):*
> ```Go
> 	// it is not set, which means tolerate the taint forever (do not evict). Zero and
> 	// negative values will be treated as 0 (tolerate the taint no time) by the system.
> 	// +optional
> ```

For Taints with effect other than NoExecute it is no-op.

---

*[pkg/api/helpers.go, line 502 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39914#-KbUTpmj9v1ufCdiPGjE:-KbUTpmj9v1ufCdiPGjF:bqtw1sn) ([raw file](https://github.com/kubernetes/kubernetes/blob/76087b742e86f2fe3af9c2bee1b4e8c9e8dbf79e/pkg/api/helpers.go#L502)):*
> ```Go
> }
> 
> func GetPodTolerations(pod *Pod) ([]Toleration, error) {
> ```

I don't follow all apimachinery changes - are those helper functions now in `v1/helpers.go`?

---

*[pkg/api/helpers.go, line 538 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39914#-KaS_lkkbbrA7ohR4kuv-r1-538:-KbUU1qmnqQlIWmcalEo:b-88sxgv) ([raw file](https://github.com/kubernetes/kubernetes/blob/76087b742e86f2fe3af9c2bee1b4e8c9e8dbf79e/pkg/api/helpers.go#L538)):*
<details><summary><i>Previously, gmarek (Marek Grabowski) wrote…</i></summary><blockquote>

Why? I can imagine use case when user want e.g. to tolerate 'NoSchedule' taint only for some time, as after that time he really need given Pod running. @davidopp 
</blockquote></details>

Will this change be automatically applied to `v1/helpers`?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39914)*

**Comment by gmarek** [comment]
@kevin-wangzefeng please rebase

**Comment by kevin-wangzefeng** [comment]
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[pkg/api/v1/helpers.go, line 340 at r6](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39914#-KbUQH1H9m7nRPP1mC_5:-Kb_kQy01eBgvudfr3y_:bla27sy) ([raw file](https://github.com/kubernetes/kubernetes/blob/bd20a10d0bfb173cb47513308266f01f8257157c/pkg/api/v1/helpers.go#L340)):*
<details><summary><i>Previously, gmarek (Marek Grabowski) wrote…</i></summary><blockquote>

since the taint was observed by the TaintManager (controller? - I don't know what's the current naming scheme)
</blockquote></details>

maybe just NodeController (regarding TaintManager as a part of NodeController)?

---

*[pkg/api/v1/helpers.go, line 351 at r6](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39914#-KbUQ_Iuxbr1y6kFKU7H:-Kb_omD09gEcZckbWMQ8:b-8a4dyp) ([raw file](https://github.com/kubernetes/kubernetes/blob/bd20a10d0bfb173cb47513308266f01f8257157c/pkg/api/v1/helpers.go#L351)):*
<details><summary><i>Previously, gmarek (Marek Grabowski) wrote…</i></summary><blockquote>

Honestly - this is annoying (especially when writing tests). Plus it's duplicated functionality of `Effect` check. We have two options here:
- keep the logic as it's currently written, which means that there will be no way of specifying with single Toleration logic:
"tolerate everything, but `NoExecute` Taints only for 10 minutes"
- remove this check (i.e. tolerate `NoSchedule` taints  even when toleration has TolerationSeconds set, the disadvantage is that we'd need to make sure that for taints other than `NoExecute` this field is ignored and make sure that all users know about this + it's easy to find.

I'm strongly supporting the latter. @davidopp @timothysc 
</blockquote></details>

Do you mean to remove the whole time-based checking here? 
When taint.timeAdded is not set, the `taint.TimeAdded.IsZero()` checking actually results the same with the TolerationSeconds calculated checking.

As TaintManager is caching taints and observing taint changes, I think it is reasonable to do time-based checking outside library (in TaintManager).

---

*[pkg/api/helpers.go, line 538 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39914#-KaS_lkkbbrA7ohR4kuv-r1-538:-Kb_gshWkd8sXrD8fghW:bt4vm63) ([raw file](https://github.com/kubernetes/kubernetes/blob/76087b742e86f2fe3af9c2bee1b4e8c9e8dbf79e/pkg/api/helpers.go#L538)):*
<details><summary><i>Previously, gmarek (Marek Grabowski) wrote…</i></summary><blockquote>

Will this change be automatically applied to `v1/helpers`?
</blockquote></details>

I don't think we need the changes here (in 'api/helpers.go') any more, just change in `v1/helpers`

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39914)*

**Comment by kevin-wangzefeng** [comment]
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[pkg/api/v1/helpers.go, line 389 at r6](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39914#-KbUSHbYDGsMMzfvXL3L:-Kb_qHZAeD4Y92fl6rDr:bhl9dbh) ([raw file](https://github.com/kubernetes/kubernetes/blob/bd20a10d0bfb173cb47513308266f01f8257157c/pkg/api/v1/helpers.go#L389)):*
<details><summary><i>Previously, gmarek (Marek Grabowski) wrote…</i></summary><blockquote>

Is this function used anywhere?
</blockquote></details>

Originally there's no filter for this func, the change is to optimize taints select (by effect) and copy operation before matching. 
Currently it is only used in scheduler.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39914)*

**Comment by kevin-wangzefeng** [comment]
Review status: 2 of 6 files reviewed at latest revision, 13 unresolved discussions.

---

*[pkg/api/v1/helpers.go, line 337 at r6](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39914#-KbUQ2_E1-JTDQheHdta:-Kb_zxkwjKeDw4iz_zyi:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/bd20a10d0bfb173cb47513308266f01f8257157c/pkg/api/v1/helpers.go#L337)):*
<details><summary><i>Previously, gmarek (Marek Grabowski) wrote…</i></summary><blockquote>

`timeAdded` doesn't impact this.  nil tolerationSeconds means that we want to tolerate given taint(s) forever.
</blockquote></details>

Done.

---

*[pkg/api/helpers.go, line 502 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39914#-KbUTpmj9v1ufCdiPGjE:-Kba-4xzaxejd-GWuWXJ:b2nhswy) ([raw file](https://github.com/kubernetes/kubernetes/blob/76087b742e86f2fe3af9c2bee1b4e8c9e8dbf79e/pkg/api/helpers.go#L502)):*
<details><summary><i>Previously, gmarek (Marek Grabowski) wrote…</i></summary><blockquote>

I don't follow all apimachinery changes - are those helper functions now in `v1/helpers.go`?
</blockquote></details>

Yes

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39914)*

**Comment by kevin-wangzefeng** [comment]
Rebased, addressed and replied to review comments, PTAL.

**Comment by gmarek** [comment]
Looks good to me. I'll let @davidopp to have one final look.

---

Reviewed 5 of 11 files at r7, 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/api/v1/helpers.go, line 351 at r6](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39914#-KbUQ_Iuxbr1y6kFKU7H:-KbewsCyx9YNEVvW0uyo:bgvl7ln) ([raw file](https://github.com/kubernetes/kubernetes/blob/bd20a10d0bfb173cb47513308266f01f8257157c/pkg/api/v1/helpers.go#L351)):*
<details><summary><i>Previously, kevin-wangzefeng (Kevin) wrote…</i></summary><blockquote>

Do you mean to remove the whole time-based checking here? 
When taint.timeAdded is not set, the `taint.TimeAdded.IsZero()` checking actually results the same with the TolerationSeconds calculated checking.

As TaintManager is caching taints and observing taint changes, I think it is reasonable to do time-based checking outside library (in TaintManager).
</blockquote></details>

Yup - my point exactly:)

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39914#-:-KbexUHeGokolcdch5ZI:bdczn49)*

File: pkg/api/v1/helpers.go
Comment (gmarek) [inline pkg/api/v1/helpers.go]: I think that we should have this check only if `taint.Effect` is `NoExecute`. I.e. 'TolerationSeconds' is a no-op for other types of taints. @davidopp @timothysc @kubernetes/sig-scheduling-pr-reviews

Comment (timothysc) [inline pkg/api/v1/helpers.go]: +1 the api comment only calls out NoExecute at this time.

File: pkg/api/v1/helpers.go
Comment (timothysc) [inline pkg/api/v1/helpers.go]: Please comment exported f(n)

**Review by timothysc — CHANGES_REQUESTED** [review]
Very minor nits

**Comment by gmarek** [comment]
@kevin-wangzefeng - please take a look.

**Comment by kevin-wangzefeng** [comment]
Review status: 4 of 7 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[pkg/api/v1/helpers.go, line 351 at r6](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39914#-KbUQ_Iuxbr1y6kFKU7H:-KbojGnnon50zV4sGkV7:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/bd20a10d0bfb173cb47513308266f01f8257157c/pkg/api/v1/helpers.go#L351)):*
<details><summary><i>Previously, gmarek (Marek Grabowski) wrote…</i></summary><blockquote>

Yup - my point exactly:)
</blockquote></details>

Done.

---

*[pkg/api/v1/helpers.go, line 333 at r8](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39914#-KaS_lkkbbrA7ohR4kv--r8-333:-KbojHVgG9WjKE-PTqAe:buq5y3w) ([raw file](https://github.com/kubernetes/kubernetes/blob/c1343d8329466570aaaebea92f3b4552c63c7fec/pkg/api/v1/helpers.go#L333)):*
Removed the whole time-based checking according to the discussion in r6 line 331 (in Reviewable).

> As TaintManager is caching taints and observing taint changes, I think it is reasonable to do time-based checking outside library (in TaintManager).

---

*[pkg/api/v1/helpers.go, line 395 at r8](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39914#-KaS_lkkbbrA7ohR4kv--r8-395:-KboioA4PY5SpHpHuO4_:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/c1343d8329466570aaaebea92f3b4552c63c7fec/pkg/api/v1/helpers.go#L395)):*
<details><summary><i>Previously, timothysc (Timothy St. Clair) wrote…</i></summary><blockquote>

Please comment exported f(n) 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/39914)*

**Comment by kevin-wangzefeng** [comment]
Refreshed, PTAL.

**Comment by kevin-wangzefeng** [comment]
Will squash after the PR is finally ready.

**Comment by gmarek** [comment]
LGTM - @timothysc please see if it's OK for you as well.

**Review by timothysc — APPROVED** [review]

**Comment by timothysc** [comment]
lgtm, @davidopp could you do final approval.

**Comment by gmarek** [comment]
@davidopp PTAL and apply the label.

**Pull request merged**
